Lib/fontTools/fontBuilder.py

```diff
@@ -0,0 +1,548 @@
+        table = self.font[tableTag] = newTable(tableTag)
+        for k, v in defaults.items():
+            setattr(table, k, v)
+        for k, v in values.items():
+            setattr(table, k, v)
+        return table
+
+    def _updateTableWithValues(self, tableTag, values):
+        table = self.font[tableTag]
+        for k, v in values.items():
+            setattr(table, k, v)
+
+    def setupHead(self, **values):
+        self._initTableWithValues("head", _headDefaults, values)
+
+    def updateHead(self, **values):
+        self._updateTableWithValues("head", values)
+
+    def setupGlyphOrder(self, glyphOrder):
+        self.font.setGlyphOrder(glyphOrder)
+
+    def setupCharacterMap(self, cmapping, allowFallback=False):
+        subTables = []
+        highestUnicode = max(cmapping)
+        if highestUnicode > 0xffff:
+            cmapping_3_1 = dict((k, v) for k, v in cmapping.items() if k < 0x10000)
+            subTable_3_10 = buildCmapSubTable(cmapping, 12, 3, 10)
+            subTables.append(subTable_3_10)
+        else:
+            cmapping_3_1 = cmapping
+        format = 4
+        subTable_3_1 = buildCmapSubTable(cmapping_3_1, format, 3, 1)
+        try:
+            subTable_3_1.compile(self.font)
+        except struct.error:
+            # format 4 overflowed, fall back to format 12
+            if not allowFallback:
+                raise ValueError("cmap format 4 subtable overflowed; sort glyph order by unicode to fix.")
+            format = 12
+            subTable_3_1 = buildCmapSubTable(cmapping_3_1, format, 3, 1)
+        subTables.append(subTable_3_1)
+        subTable_0_3 = buildCmapSubTable(cmapping_3_1, format, 0, 3)
+        subTables.append(subTable_0_3)
+
+        self.font["cmap"] = newTable("cmap")
+        self.font["cmap"].tableVersion = 0
+        self.font["cmap"].tables = subTables
+
+    def setupNameTable(self, nameStrings):
+        nameTable = self.font["name"] = newTable("name")
+        nameTable.names = []
+
+        for nameName, nameValue in nameStrings.items():
+            if isinstance(nameName, int):
+                nameID = nameName
+            else:
+                nameID = _nameIDs[nameName]
+            if isinstance(nameValue, basestring):
+                nameValue = dict(en=nameValue)
+            nameTable.addMultilingualName(nameValue, ttFont=self.font, nameID=nameID)
+
+    def setupOS2(self, **values):
+        if "xAvgCharWidth" not in values:
+            gs = self.font.getGlyphSet()
+            widths = [gs[glyphName].width for glyphName in gs.keys() if gs[glyphName].width > 0]
+            values["xAvgCharWidth"] = int(round(sum(widths) / float(len(widths))))
+        self._initTableWithValues("OS/2", _OS2Defaults, values)
+        if not ("ulUnicodeRange1" in values or "ulUnicodeRange2" in values or
+                "ulUnicodeRange3" in values or "ulUnicodeRange3" in values):
+            assert "cmap" in self.font
```

Comment:
it would be good to document any of these inter-table dependencies, so that it's clear what's the expected order that one may call the various setupTable methods.